pkg/api/query/grpc.go

```diff
@@ -115,7 +115,14 @@ func (g *GRPCAPI) Query(request *querypb.QueryRequest, server querypb.Query_Quer
 	default:
 		return status.Error(codes.InvalidArgument, "invalid engine parameter")
 	}
-	qry, err := engine.NewInstantQuery(ctx, queryable, promql.NewPrometheusQueryOpts(false, lookbackDelta), request.Query, ts)
+
+	var queryExpr string
+	if request.QueryPlan != nil {
+		queryExpr = request.QueryPlan.String()
+	} else {
+		queryExpr = request.Query
+	}
+	qry, err := engine.NewInstantQuery(ctx, queryable, promql.NewPrometheusQueryOpts(false, lookbackDelta), queryExpr, ts)
```

Comment:
we are kind discarding the benefit from having the plan in the request again here; i think we should use the constructor that uses the plan here!

Reply:
Is it okay if I do a safe cast to check if we are actually using the engine.DistributedEngine? Since the method is not part of the interface `promql.QueryEngine`.

Reply:
Check the last version please